staging/src/k8s.io/kube-aggregator/pkg/apiserver/apiserver.go

```diff
@@ -243,7 +248,8 @@ func (s *APIAggregator) AddAPIService(apiService *apiregistration.APIService) er
 	// since they are wired against listers because they require multiple resources to respond
 	if proxyHandler, exists := s.proxyHandlers[apiService.Name]; exists {
 		proxyHandler.updateAPIService(apiService)
-		return s.openAPIAggregator.loadApiServiceSpec(proxyHandler, apiService)
+		s.openAPIAggregationController.UpdateAPIService(proxyHandler, apiService)
```

Comment:
reading top to bottom, this surprised me since I would have expected the openapiaggregation controller to drive based on its own sharedinformer and only aggregate things that are `Available` and un-aggregate them when they are not available.  That would avoid coupling the bits, right?

Reply:
Good idea :+1:

Reply:
How about this?
1- OpenAPISpecStore (implements store functionality)
2- Controller keeps the Queue and Key structure it has, and Add/Update/Delete Store
3- OpenAPI Aggregator watch OpenAPISpecStore and aggregate changes

Reply:
not sure what you mean with the store here. The basic pattern is that a controller uses a shared informer. This decouples the controllers while sharing a connection to the apiserver. So it's just a matter of plumbing your controller with its own shared informer. Should be pretty straight forward. Just look at some of the many examples.

In your control loop you can filter by the Available condition.

Reply:
I will look at some examples to understand this.

Reply:
> sharing a connection to the apiserver
That would apply to informers on rest endpoints/api objects.

Is there an example of custom informer? i.e. informer for a non-api object?

Reply:
They only exist for API resources.

Reply:
@deads2k I've discussed this in details with @sttts. Thought it looks easy to do at the beginning, it is quite complex to do. Right now OpenAPI controller gets its signal when Add/Remove APIService in APIAggregator is being called. That is all it needs to function. But there is quite a logic in APIServiceRegistrationController that listen on two informers, combine their signals with some logic and calls AddAPIService and RemoveAPIService. If we want to decouple OpenAPI controller from those two functions we end up either duplicating the code in APIServiceRegistrationController or factor them out somehow to be used in both controllers or have OpenAPIAggregationController derive from APIServiceRegistrationController and changes the queue and sync function (and log messages, etc.). All of these solutions are quite complex and we think it does not worth to do, unless we missed something here. would you agree that this decoupling is not crucial to this PR and current solutions are quite complex that deserve their own time to fix with a new PR (if we found a simple solution to it)?

Reply:
not to forget that we also need access to the `ProxyHandler`. So we get a couple of new (code) dependencies in exchange of the three add/remove/update callbacks. I would prefer the later.

Reply:
> would you agree that this decoupling is not crucial to this PR and current solutions are quite complex that deserve their own time to fix with a new PR (if we found a simple solution to it)?

Thanks for looking at it.  At some point we should restructure in a way that makes it more reasonable, but I'm ok defering as long as this is non-blocking as it is.  My recollection is that is already done.

Reply:
Yes, it is non-blocking. Thanks.